config/store.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package config
+
+import (
+	"github.com/mattermost/mattermost-server/model"
+)
+
+// Listener is a callback function invoked when the configuration changes.
+type Listener func(oldConfig *model.Config, newConfig *model.Config)
+
+// Store abstracts the act of getting and setting the configuration.
+type Store interface {
+	// Get fetches the current, cached configuration.
+	Get() *model.Config
+
+	// GetEnvironmentOverrides fetches the configuration fields overridden by environment variables.
+	GetEnvironmentOverrides() map[string]interface{}
+
+	// Set replaces the current configuration in its entirety, without updating the backing store.
+	Set(*model.Config) (*model.Config, error)
+
+	// Patch merges the given config with the current configuration, without updating the backing store.
```

Comment:
The thought just occurred that it might make sense to have `Set` have `Patch` semantics for any `nil` fields as such. 0/5, but that would eliminate another method, at the expense of perhaps being less clear (unless we renamed `Set` to `Merge`?)